mcumgr-smp/src/transport/ble/mod.rs

```
@@ -0,0 +1,115 @@
+            peripheral_device.ok_or(Error::BLE(btleplug::Error::DeviceNotFound))?;
+
+        peripheral_device.connect().await?;
+        peripheral_device.discover_services().await?;
+        let smp_char = peripheral_device
+            .characteristics()
+            .into_iter()
+            .find(|attr| attr.uuid == SMP_CHAR)
+            .ok_or(Error::BLE(btleplug::Error::NoSuchCharacteristic))?;
+
+        peripheral_device.subscribe(&smp_char).await?;
+
+        let notifications = peripheral_device.notifications().await?;
+
+        Ok(Self {
+            peripheral_device,
+            notifications,
+            smp_char,
+        })
+    }
+
+    pub async fn from_peripheral(device: Peripheral) -> Result<Self, Error> {
+        device.connect().await?;
+        device.discover_services().await?;
+        let smp_char = device
+            .characteristics()
+            .into_iter()
+            .find(|attr| attr.uuid == SMP_CHAR)
+            .ok_or(Error::BLE(btleplug::Error::NoSuchCharacteristic))?;
+
+        device.subscribe(&smp_char).await?;
+
+        let notifications = device.notifications().await?;
+
+        Ok(Self {
+            peripheral_device: device,
+            notifications,
+            smp_char,
+        })
+    }
+}
+
+#[async_trait]
+impl SmpTransportAsync for BleTransport {
+    async fn send(&mut self, frame: Vec<u8>) -> Result<(), Error> {
+        self.peripheral_device
+            .write(
+                &self.smp_char,
+                &frame,
+                btleplug::api::WriteType::WithoutResponse,
+            )
+            .await?;
+        Ok(())
+    }
+
+    async fn receive(&mut self) -> Result<Vec<u8>, Error> {
+        loop {
+            if let Some(res) = self
+                .notifications
+                .next()
+                .await
+                .filter(|notif| notif.uuid == SMP_CHAR)
+            {
+                return Ok(res.value);
+            }
+        }
```

if the stream returns None (for whatever reason) you should return an error from the loop.